## a, b)

### Definindo o diretório de trabalho e importando os dados e bibliotecas necessárias.

In [11]:
import os
import pandas as pd
import chardet
import plotly.express as px

# caminho do diretório
diretorio = r"C:\Users\luisf\OneDrive\Desktop\DATACAMP\FGV_IBRE"

# Altera o diretório
os.chdir(diretorio)


# Leitura do arquivo para detecção do encoding
with open('base_ipea.csv', 'rb') as f:
    dados = f.read() 

resultado = chardet.detect(dados)

encoding_detectado = resultado['encoding']

    
# encontrei o enconding utilizando o notepad++ apesar de poder ter feito utilizando a bibliotecas chardet
base_ipea = pd.read_csv('base_ipea.csv', encoding= encoding_detectado, sep=';')
base_ipea.head()

região onde foi realizada a entrevista   
0                                  NORTE  \
1                                  NORTE   
2                                  NORTE   
3                               NORDESTE   
4                               NORDESTE   

   idade: qual a sua idade (em anos completos)?    área       sexo   
0                                            34  urbano   feminino  \
1                                            53  urbano  masculino   
2                                            31  urbano  masculino   
3                                            38  urbano   feminino   
4                                            21  urbano   feminino   

  cor ou raça última série escolar que concluiu com aprovação   
0       parda                        quinta série fundamental  \
1      branca                                  médio completo   
2       preta                     analfabeto ou sem instrução   
3       parda                        sétima série fundamental   
4       parda                            primeira série médio   

       até que série o seu pai estudou      até que série a sua mãe estudou   
0          analfabeto ou sem instrução    quarta série fundamental completa  \
1  oitava série fundamental incompleta  oitava série fundamental incompleta   
2          analfabeto ou sem instrução          analfabeto ou sem instrução   
3         cursou superior sem concluir                  segunda série médio   
4          analfabeto ou sem instrução          analfabeto ou sem instrução   

   renda total do chefe da família no último mês   
0                                         1000.0  \
1                                         4800.0   
2                                          678.0   
3                                         1000.0   
4                                          678.0   

   renda total de todos os moradores (parentes e agregados) no último mês   
0                                             1490.0                       \
1                                             4900.0                        
2                                              678.0                        
3                                             1000.0                        
4                                              678.0                        

   ...  a roupa suja deve ser lavada em casa   
0  ...                        concorda total  \
1  ...                        concorda total   
2  ...                        discorda total   
3  ...                        concorda total   
4  ...                      concorda parcial   

   uma mulher só se sente realizada quando tem filhos/as   
0                                     discorda total      \
1                                     discorda total       
2                                                 ns       
3                                     concorda total       
4                                   discorda parcial       

  a mulher casada deve satisfazer o marido na cama, mesmo quando não tem vontade   
0                                     discorda total                              \
1                                     discorda total                               
2                                     discorda total                               
3                                   discorda parcial                               
4                                   discorda parcial                               

  piada de preto é só brincadeira, não é racismo   
0                               concorda parcial  \
1                               concorda parcial   
2                               concorda parcial   
3                               discorda parcial   
4                                 concorda total   

  um casal de dois homens vive um amor tão bonito quanto entre um homem e uma mulh   
0                                     concorda total                                \
1                                 

## c)

### Dimensões da base de dados

In [12]:
base_ipea.shape

(3810, 40)

Como visto acima utilizando o método .shape é de 3810 linhas e 40 colunas.

## d, e)

### Frequência de pessoas por região e região mais frequênte

In [56]:
df_freq = base_ipea['região onde foi realizada a entrevista'].value_counts().reset_index()
df_freq = df_freq.rename(columns={'região onde foi realizada a entrevista':"regiao", 'count':'frequencia'})
df_freq

regiao  frequencia
0       SUDESTE        1605
1      NORDESTE        1065
2           SUL         555
3         NORTE         300
4  CENTRO-OESTE         285

In [14]:
# Gráfico das frequências.
fig = px.bar(data_frame=df_freq, x = 'regiao', y = 'frequencia', color='regiao', title="Frequência de Pessoas por Região",
             labels={'regiao':'Região', 'frequencia':'Frequência'})

fig.show()

Com base nas saídas acima, podemos ver que a região mais frequente e a que é mais frequêntada por pessoas é a região Sudeste.

## f, g)

### Média Amostral, Moda, Mediana, Mínimo e Máximo da variável `idade`.

In [15]:
# encontra a idade máxima e mínima e calcula a média, mediana e moda.
def medidas(df, incluir_qt = False, incluir_moda=True):
    """ Calcula as medidas necessárias de uma variável númerica"""
    min = df.min()
    max = df.max()
    media = df.mean()
    mediana = df.median()
    moda = df.mode()[0]
    q1 = df.quantile(0.25)
    q3 = df.quantile(0.75)
    
    if incluir_moda == True:
        print(f' Mínimo: {min} \n Máximo: {max} \n Média: {media} \n Mediana(): {mediana} \n Moda: {moda}')
    elif incluir_qt == True:
        print(f' Mínimo: {min} \n Máximo: {max} \n Média: {media} \n Mediana: {mediana} \n Primeiro Quartil: {q1} \n Terceiro Quartil: {q3}')  
    else:
       print(f' Mínimo: {min} \n Máximo: {max} \n Média: {media} \n Mediana: {mediana} \n Primeiro Quartil: {q1} \n Terceiro Quartil: {q3} \n Moda: {moda}')         
    
medidas(base_ipea['idade: qual a sua idade (em anos completos)?'])    

 Mínimo: 16 
 Máximo: 88 
 Média: 42.31784776902887 
 Mediana(): 41.0 
 Moda: 18


Como podemos ver, a idade mínima, e máxima são respectivamente 16 anos e 88 anos. Além disso, as medidas de centralidade implicam que a variável `idade` tem distribuição assimétrica negativa, dado que a moda é menor que a média e mediana.

## h)

In [16]:
faixa_etaria = ["Jovens", "Adultos", "Idosos"]
faixa_etaria_limites = [0 , 29, 59, 89]

base_ipea['faixa_etaria'] = pd.cut(base_ipea['idade: qual a sua idade (em anos completos)?'], bins=faixa_etaria_limites, labels=faixa_etaria)
base_ipea[['faixa_etaria', 'idade: qual a sua idade (em anos completos)?']]

faixa_etaria  idade: qual a sua idade (em anos completos)?
0         Adultos                                            34
1         Adultos                                            53
2         Adultos                                            31
3         Adultos                                            38
4          Jovens                                            21
...           ...                                           ...
3805      Adultos                                            39
3806      Adultos                                            40
3807      Adultos                                            37
3808       Jovens                                            19
3809      Adultos                                            38

[3810 rows x 2 columns]

In [57]:
base_ipea['faixa_etaria'].value_counts().reset_index().rename(columns={'count':'frequência'})

faixa_etaria  frequência
0      Adultos        1995
1       Jovens        1087
2       Idosos         728

In [18]:
px.bar(data_frame=base_ipea['faixa_etaria'].value_counts().reset_index(), x='faixa_etaria', y='count', color='faixa_etaria', title='Número de Pessoas por Faixa Etária',
       labels={'faixa_etaria':'Faixa Etária', 'count':'Frequência'})

Conforme as saídas acima, verificamos que a maioria da amostra é composta por Adultos.

# i)

In [19]:
# calcula as medidas conforme o enunciado, eu poderia ter usado o .describe() aqui, já que ele calcula todas as medidas pedidas
# porém como a função medidas() ja estava praticamente feita para esse fim relsovi utiliza-la para manter o codigo mais conciso

medidas(base_ipea['renda total de todos os moradores (parentes e agregados) no último mês'], incluir_moda=False, incluir_qt=True)

 Mínimo: 150.0 
 Máximo: 19000.0 
 Média: 1801.4782211392019 
 Mediana: 1400.0 
 Primeiro Quartil: 800.0 
 Terceiro Quartil: 2150.0



O interessante dessas medidas, é que se utilizarmos um box-plot, ou calcularmos os outliers utilizando: $Q1 - (1.5*IQR)$ para os outliers inferiores, e $Q3 - (1.5*IQR)$ para os outliers superiores, em que o $IQR=Q3-Q1$ é a amplitute interquartil, encontramos que o valor mínimo R$ 150,00 não é considerado como outlier, mas o valor máximo de R$ 19.000,00 é considerado, isso pode indicar desigualdade de renda, além disso, a mediana da renda total de todos os moradores (parentes e agregados) no último mês é de R$ 1.400,00, o que nos ajuda a reforçar a hipótese de que pode existir desigualdade social nesta amostra. Vale ressaltar que a mediana é uma medida mais confiável aqui, pois a média é mais sensível aos outliers com valores extremos.  
<br>
<br>

##  j)
O que podemos análisar sobre o primeiro e terceiro quartil, vai de encontro com o que foi apontado acima. O primeiro quartil indíca que 25% das rendas registradas estão abaixo de R$ 800,00 e 75% estão acima deste valor, ao ponto que pelo terceiro quartil 75% das rendas registradas estão abaixo de R$ 2.150,00 e 25% estão acima deste valor, o que ajuda reforçar a hipótese de desigualdade social.


# l)

### Coeficiente de Variação

In [32]:
def coeficiente_variacao(df):
    """
    Calcula o coeficente de variação
    """   
       
    if len(df) < 2:
        return None

    media = df.mean()
    desvio_padrao = df.std()

    cv = (desvio_padrao / media)
    return cv

cv_idade = coeficiente_variacao(base_ipea['idade: qual a sua idade (em anos completos)?'])
cv_renda = coeficiente_variacao(base_ipea['renda total de todos os moradores (parentes e agregados) no último mês'])

print(f'O coeficiente de variação para a variável idade é: {cv_idade} e para variável renda é: {cv_renda}')

O coeficiente de variação para a variável idade é: 0.399740573585868 e para variável renda é: 0.8551213791858479


Para variável idade o coeficiente de variação é relativamente baixo o que nos diz que os valores valores de idade estão próximos da média, com uma dispersão relativamente baixa em relação à a mesma. Portanto, existe uma menor variabilidade na variável idade. 

Já para renda, o oposto ocorre, os valores de renda estão mais dispersos em relação à média, apresentando assim, uma maior variabilidade. Portanto, podemos dizer que os dados de renda possuem uma dispersão maior em relação à média, indicando uma variabilidade mais elevada, o que era de se esperar com base na análise da renda.

## m) Calcule o desvio-padrão para a renda de acordo com cada região do Brasil

In [54]:
std_df = base_ipea.groupby('região onde foi realizada a entrevista')['renda total de todos os moradores (parentes e agregados) no último mês'].agg('std').reset_index()
std_df = std_df.rename(columns={'região onde foi realizada a entrevista':'Região','renda total de todos os moradores (parentes e agregados) no último mês':'Desvio Padrão'})
std_df

Região  Desvio Padrão
0  CENTRO-OESTE    1343.430396
1      NORDESTE    1189.550086
2         NORTE    1757.768796
3       SUDESTE    1670.977045
4           SUL    1316.884815

Como vemos na saída, as pessoas da região Nordeste possuem renda com um comportamento mais homogêneo, dado que o desvio padrão é o menor entre as regiões.